compiler/rustc_data_structures/src/sync.rs

```diff
@@ -460,6 +463,91 @@ impl<T: Clone> Clone for Lock<T> {
     }
 }
 
+fn next() -> NonZeroUsize {
+    static COUNTER: std::sync::atomic::AtomicUsize = std::sync::atomic::AtomicUsize::new(1);
+    NonZeroUsize::new(COUNTER.fetch_add(1, Ordering::SeqCst)).expect("more than usize::MAX threads")
+}
+
+pub(crate) fn get_thread_id() -> NonZeroUsize {
+    thread_local!(static THREAD_ID: NonZeroUsize = next());
+    THREAD_ID.with(|&x| x)
+}
```

Comment:
Don't we have something in `std::thread` or in `rayon` for this?

Reply:
Yea, it looks like we can use `std::thread::ThreadId` here